illumos-utils/src/contract.rs

```diff
@@ -0,0 +1,375 @@
+    }
+}
+
+impl Watcher {
+    /// Return a Watcher for a specific type of contract event.  The watcher
+    /// will return all events of the requested type, and it is the caller's
+    /// responsibility to filter for the events relevent to them.
+    pub fn new(typ: ContractType) -> Self {
+        let path = path(typ, None, "pbundle");
+        let fd = unsafe { libc::open(path.as_ptr(), libc::O_RDONLY) };
+        if fd < 0 {
+            panic!(
+                "Could not open {:?}: {}",
+                path,
+                std::io::Error::last_os_error()
+            );
+        }
+        Watcher { fd }
+    }
+
+    /// Block until a contract event occurs.
+    pub fn watch(&self, log: &slog::Logger) -> ContractEvent {
+        loop {
+            let mut event: ct_evthdl_t = std::ptr::null_mut();
+            let evp: *mut ct_evthdl_t = &mut event;
+            // The event endpoint was not opened as non-blocking, so
+            // ct_event_read_critical(3CONTRACT) will block until a new
+            // critical event is available on the channel.
+            match unsafe { ct_event_read_critical(self.fd, evp) } {
+                0 => {
+                    let typ = unsafe { ct_event_get_type(event) };
+                    let event_id = unsafe { ct_event_get_evid(event) };
+                    let ctid = unsafe { ct_event_get_ctid(event) };
+                    return ContractEvent { event, event_id, typ, ctid };
+                }
+                err => {
+                    // ct_event_read_critical(3CONTRACT) does not state any
+                    // error values for this function if the file descriptor
+                    // was not opened non-blocking, but inspection of the
+                    // library code shows that various errnos could be returned
+                    // in situations such as failure to allocate memory. In
+                    // those cases, log a message and pause to avoid entering a
+                    // tight loop if the problem persists.
+                    error!(
+                        log,
+                        "Unexpected response from contract event channel: {}",
+                        std::io::Error::from_raw_os_error(err)
+                    );
+                    thread::sleep(Duration::from_secs(1));
+                }
+            }
+        }
+    }
+}
+
+/// A Control is used to communicate a response to the contract system in
+/// response to an event.  In practice this is limited to acknowledging an event
+/// and cancelling a contract.
+pub struct Control {
+    ctid: ctid_t,
+    fd: c_int,
+}
+
+impl Drop for Control {
+    fn drop(&mut self) {
+        unsafe { libc::close(self.fd) };
+    }
+}
+
+impl Control {
+    /// Construct a new Control for the specified contract
+    pub fn new(
+        typ: ContractType,
+        ctid: ctid_t,
+    ) -> Result<Self, crate::ExecutionError> {
+        let path = path(typ, Some(ctid), "ctl");
```

Comment:
It looks like when this is called by `process_contract_reaper`, this constructs the path

```
format!("/system/contract/process/{}/ctl", ctid)
```

but the old `abandon_contract()` function constructed the path

```
format!("/proc/{}/contracts/{}/ctl", process::id(), ctid)
```

Are those equivalent?

This module has enough rework of the contract wrapping (and has a fair bit of `unsafe`) that I'd love to get a couple different reviews on it - maybe @citrus-it could take a look too? I'm happy to review from a purely FFI point of view, but I'm having a hard time comparing the old and new when it's in a different file and has a bunch of structural changes, so am a lot less confident nothing changed accidentally.